packages/@aws-cdk/aws-batch/lib/compute-environment.ts

```diff
@@ -0,0 +1,432 @@
+            type: props ? (props.type ? props.type : ComputeEnvironmentType.MANAGED) : ComputeEnvironmentType.MANAGED,
+        });
+
+        this.computeEnvironmentArn = computeEnvironment.ref;
+        this.computeEnvironmentName = this.getResourceNameAttribute(props ? props.computeEnvironmentName || this.physicalName : this.physicalName);
+    }
+
+    /**
+     * Validates the properties provided for a new batch compute environment
+     */
+    public validateProps(props?: ComputeEnvironmentProps) {
+        if (props === undefined) {
+            return;
+        }
+
+        // Setting a bid percentage is only allowed on SPOT resources
+        if (props.computeResources && props.computeResources.type &&
+            props.computeResources.type === ComputeResourceType.EC2 && props.computeResources.bidPercentage !== undefined) {
+            throw new Error('Setting the bid percentage is only allowed for SPOT type resources on a batch compute environment');
+        }
+
+        if (props.computeResources && props.computeResources.type && props.allocationStrategy &&
+            props.computeResources.type === ComputeResourceType.EC2 && props.allocationStrategy === AllocationStrategy.SPOT_CAPACITY_OPTIMIZED) {
+                throw new Error('The SPOT_CAPACITY_OPTIMIZED allocation strategy is only allowed if the environment is a SPOT type compute environment');
+        }
+
+        // Bid percentage must be from 0 - 100
+        if (props.computeResources && props.computeResources.bidPercentage !== undefined &&
+            (props.computeResources.bidPercentage < 0 || props.computeResources.bidPercentage > 100)) {
+            throw new Error('Bid percentage can only be a value between 0 and 100');
+        }
+
+        if (props.computeResources && props.computeResources.minvCpus &&
+            props.computeResources.minvCpus < 0) {
+            throw new Error('Minimum vCpus for a batch compute environment cannot be less than 0');
+        }
+
+        if (props.computeResources && props.computeResources.minvCpus && props.computeResources.maxvCpus &&
+            props.computeResources.minvCpus > props.computeResources.maxvCpus) {
+            throw new Error('Minimum vCpus cannot be greater than the maximum vCpus');
+        }
+    }
+
+    private buildInstanceTypes(instanceTypes?: ec2.InstanceType[]): string[] {
+        if (instanceTypes === undefined) {
+            return [
+                'optimal',
+            ];
+        }
+
+        return instanceTypes.reduce((types: string[], type: ec2.InstanceType): string[] => {
+            return [...types, type.toString()];
+        }, []);
+    }
+
+    private buildSecurityGroupIds(securityGroupIds?: ec2.ISecurityGroup[]): string[] | undefined {
+        if (securityGroupIds === undefined) {
+            return undefined;
+        }
+
+        return securityGroupIds.reduce((ids: string[], group: ec2.ISecurityGroup): string[] => {
```

Comment:
Same comment here (should be `map`, not `reduce`).

Reply:
Got it 👍